### DISCUSSION TITLE
service setup

### DISCUSSION THREAD

**Discussion by edward3h** [alex-konshin/f007th-rpi#9] [General]
Hi,
Just wanted to let you know, I've been working on a script and unit files to set it up as a systemd service. Not that close to done yet though.

**Comment by alex-konshin** [comment]
Actually this is in my plans but I am not sure that you can see it: https://github.com/alex-konshin/f007th-rpi/projects/1

I am going to work on it before New Year. I am on vacation at my home and have some free time. I do not expect it to be difficult but I first need to do the same for gpio-ts otherwise the utility will not be able to start automatically. I know that it is possible to register kernel module to be auto loaded during boot, I just have not search for it and have not documented it yet. 

The major problem for me is documentation because I am not very fluent in English. It is very time-consuming. GitHub markdown is also not very convenient and requires many tries to get acceptable results. I know that GitHub wiki also supports MediaWiki and I have tried it too but for me it works even worse than markdown.

**Reply by alex-konshin** [reply]
I have done part of the task - I finally made gpio-ts module to be loaded during boot. It was not easy to find the right way to do it. I tried many different variants to do it before I found a working solution.\
One of culprits was this: the name of the file is `gpio-ts.ko` but the name of the module is `gpio_ts` (note dash and underscore in names).

All these steps have to be done under root account (or via sudo).
- Under root account (or via sudo) copy file `gpio-ts.ko` to `/lib/modules/\`uname -r\`/kernel/drivers/gpio/gpio-ts.ko`
  Actually I changed the ownership of this new file to `root:root` but I am not sure that it is really required.
- Run command `depmod`.
- Create file `/etc/modprobe.d/gpio_ts.conf` with following content:
  ```
  options gpio_ts gpios=27
  ```
  Note that the name of the module is `gpio_ts` rather than `gpio-ts`.
- Create file `/etc/modules-load.d/gpio_ts.conf` with following content:
  ```
  gpio_ts
  ```
  Actually adding this line in /etc/modules should also have the same effect but for unknown reason it did not.\
  One more note: you cannot put module parameters in this file now. Parameters can be specified in a file in directory /etc/modprobe.d/.
- Reboot with command `reboot`.
- After booting check that the module is really loaded: `lsmod | grep gpio_ts`

Now I need to document it (well, it is partially done above) and check other platforms. Probably it will be different because it depends on kernel versions. Then I will try to create a script to automate these steps.

I think the part with defining service for f007th-send will be easier. I will look at it tomorrow.

**Comment by edward3h** [comment]
This is what I hacked up on my machine to get a service running for now. Since the module is connected to f007th-send by having to use the same GPIO number, I have a wrapper script do both things.
https://gist.github.com/edward3h/40f1d9a85c8ad34c05750940e0de1d45

**Reply by alex-konshin** [reply]
Well, I have already taken care of automatic loading of gpio-ts during boot. I submitted script set_load_on_boot_rpi.sh to gpio-ts project before New Year. And I even tested it on both Raspberry Pi 4 (Raspbian) and MinnowBoard (Ubuntu). Actually I also tested it on RPi 3 but I have upgraded OS to buster before it (and it was not streight forward) and, as result, I cannot claim that it works with older versions of Linux. Anyway it is possible to setup auto loading of gpio-ts and it is relatively easy task.
So, we don't need to deal with gpio-ts in these service scripts. We can but I don't see much value.
